### ISSUE TITLE
Peg-Out Clarity Contract for Request & Fullfilment

### ISSUE THREAD

**Issue by empeje** [Trust-Machines/core-eng#157]
## Background

Taking directly from the paper

> This is the project encompassing the work required to enable the sBTC signers to process a peg-in.  It must satisfy the following user stories:
> - As the Stackers, we can get informed by a user of the spending script that matches a Bitcoin transaction which funded a user’s taproot address.  This transaction might not be in a Bitcoin block; it might be in the mempool.
> - As the set of Stackers, we can spend the user’s taproot address and complete the peg-in as long as the user-specified timeout in their spending condition has not passed.
> - As the Stackers, we can opt to ignore the user’s peg-in request for any reason.  For example, this can be to satisfy OFAC requirements, or to avoid taking on too much BTC, or to avoid taking BTC when the peg is in a bad state.
> - As a vigilante, I can post a Stacker-signed peg-out transaction to `sbtc` to materialize the peg-out request in Clarity.
> - As an observer, I can see how many peg-out requests are currently unsatisfied in `.sbtc`, including how much they are worth, how long they have been pending, and to which addresses the BTC will be sent.
> - As a vigilante, I can call into the `sbtc` contract to transition the peg to the bad state.  This will re-stack the Stackers’ STX to a burn address for the next reward cycle if peg-out requests have been pending for too long, and prevent subsequent peg-ins from materializing.  This can occur indirectly as part of another API call.
> - As a vigilante, I can post a Stacker-signed peg-out fulfillment transaction to `sbtc` to satisfy a peg-out request.  This request does not yet need to exist in `sbtc`.
> - As an observer, I can see which peg-out requests have been satisfied in `sbtc`, and I can also see “preemptively satisfied” peg-out fulfillments for peg-out requests that have not yet materialized to `sbtc`.
> - As a vigilante, I can call into the `sbtc` contract to transition the peg to the good state if there are no more too-stale pending peg-out requests.  This will permit peg-in transaction relays to resume.  This can happen indirectly as part of another API call.

The peg-out process, like pegging-in, starts on the Bitcoin blockchain. A user will generate a P2TR address with a script that allows the Stackers to spend it, while at the same time revealing a user signature used to link the user to a Stacks account. The user funds the address with some Bitcoin to pay for the fees of pegging out. That transaction and its unlock script are submitted to the sBTC protocol, locking the user’s sBTC for a predetermined duration. The Stackers then spend that output to pay for the transaction fee of fulfilling the peg-out request. If the Stackers do not process the peg-out in time, then it will expire, allowing the user to reclaim the locked sBTC tokens.

## Definition
The contract can receive and process both peg-out request transactions (peg-out fee commitments) as well as peg-out fulfilment transactions. (Stackers paying the user.) It can also process reclaim transactions that will unlock previously locked sBTC tokens.

## Motivation
This contract processes one of the two main mechanisms of the sBTC protocol. People cannot trustlessly convert their sBTC back to native Bitcoin without it.

## Technical Risk

There are currently three unresolved protocol design issues, the resolution of which will inform the design of the peg-out processing contract.

- Peg-out payload replay attack: the signature payload in the peg-out request transaction is currently calculated over only two inputs: the total BTC being pegged out and the user’s destination scriptPubHash. It means that the payload is technically a forever-approval of that particular peg-out. Anyone will be able to publish peg-outs with the same payload, which will then trigger further peg-outs for the original sender until their sBTC is (almost) fully pegged out.
- Unlock script verification issue: the unlock script is provided together with the mined peg-out request fee commitment transaction. It is doable to verify the Bitcoin transaction was mined but it is difficult (or maybe even impossible) to verify that the supplied unlock script can actually spend the output. The witness program is a 32 byte tweaked public key that is hard to calculate in Clarity.
- The contract cannot work without the ability to check if a SegWit transaction was mined. It also needs to be able to retrieve the principal from the witness data.

It would also be important to be supplied with real test vectors by the Bitcoin transaction WG.

## Dependencies
Clarity-bitcoin library with SegWit support.
Depending on the final protocol design: ability to verify a Taproot witness program.

## Rough implementation plan & timeline
Completion is measured in five sub-deliverables:

- It can receive Bitcoin peg-out request fee commitment transactions and unlock scripts, extract the necessary data, and lock user sBTC tokens.
- It can verify the unlock script is in the expected format.
- It can verify the unlock script can actually unlock the peg-out request fee commitment.
- It can receive Bitcoin peg-out fulfilment transactions as produced by the Stackers and complete the peg-out, burning the locked tokens of (1).
- It can receive reclaim transactions as produced by the user, unlocking the sBTC tokens locked in (1).
- The workload estimate for this project is 14 workdays. Although the unlock script verification issue as described in the technical risk may cause significant delay.

**Comment by jo-tm** [comment]
- General logic is there.
- TX decoding depends on new library, to be tested.
- Request ok.
- Fullfilment ok.
- Reclaim function added, new mechanism.
- MERGED into main.

**Comment by jo-tm** [comment]
- Ask Muneeb if the Non-Voting Stacking Proofs can be added on the following iteration or not.
- Still waiting final structure for the TX coming from Blockchain WG and UX WG. Ping them.

**Comment by jo-tm** [comment]
@netrome I think there was some debate on this in the Design Meeting today. Do we have a final Bitcoin TX structure to be checked in Clarity Peg-Out?

**Comment by netrome** [comment]
> @netrome I think there was some debate on this in the Design Meeting today. Do we have a final Bitcoin TX structure to be checked in Clarity Peg-Out?

I would expect there only to be minor changes coming to the structure of the peg out transactions. The main issues remaining to be addressed are the peg out request payload replayability, and how to link peg out requests and fulfillment in a many-to-many fashion.

@soju-drinker is working on our test vector scripts, but meanwhile here are some example serialized transactions for the peg out request:

Using OP_RETURN:
02000000010000000000000000000000000000000000000000000000000000000000000000ffffffff00ffffffff0300000000000000004f6a4c4c69643e000000000000053900dc18d08e2ee9f476a89c4c195edd402610176bb6264ec56f3f9e42e7386c543846e09282b6f03495c663c8509df7c97ffbcd2adc537bbabe23abd828a52bc8cd390500000000000022512000000000000000000000000000000000000000000000000000000000000000002a00000000000000225120000000000000000000000000000000000000000000000000000000000000000000000000

Using OP_DROP in p2tr-over-p2sh
020000000001010000000000000000000000000000000000000000000000000000000000000000ffffffff23225120273cb9363da5933fe6f2bdcc3fd27f06865227cb526641cc6dc47de5e9ee06edffffffff030000000000000000056a03696477390500000000000022512000000000000000000000000000000000000000000000000000000000000000002a000000000000002251200000000000000000000000000000000000000000000000000000000000000000027e4c563e000000000000053900dc18d08e2ee9f476a89c4c195edd402610176bb6264ec56f3f9e42e7386c543846e09282b6f03495c663c8509df7c97ffbcd2adc537bbabe23abd828a52bc8cddeadbeef000000000000002a7576a9200c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c87ad613c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c00000000

**Comment by netrome** [comment]
> * @netrome both TXs are outdated.
> 
>     * contract only has scafolding, will be completed after Peg-In.

Outdated in the sense that you need commit-reveal transactions? I'll ping you once they're ready.

**Comment by jo-tm** [comment]
no updates.

**Comment by jo-tm** [comment]
- will update on this after Peg-in

**Comment by jo-tm** [comment]
@MarvinJanssen can we start working on this task this week or which date you estimate?

**Comment by MarvinJanssen** [comment]
This week.

**Comment by jo-tm** [comment]
@MarvinJanssen new PR will advance this Task.
There is not blocked by #275 but is related.

**Comment by MarvinJanssen** [comment]
@MarvinJanssen started on it. Early stages, will use script verification logic from #529.

**Comment by jo-tm** [comment]
Depends on how quick we can make Bitcoin test cases using the Clarity Bitcoin library.

**Comment by jo-tm** [comment]
Please update on this issue @MarvinJanssen .

**Comment by MarvinJanssen** [comment]
I am working on peg-out redeem transaction verification, as well as a concept of STX side sponsored transaction peg out requests. (Easier to explain it with a bit of code than doing a write up. Then we can decide if we want to include this path or not.)
